sycamore/llms/openai.py

```diff
@@ -0,0 +1,252 @@
+            # reliably.
+            if model.is_chat:
+                cls = AzureOpenAIChat
+            else:
+                cls = AzureOpenAIInstruct
+
+            return cls(
+                model=model.name,
+                azure_endpoint=self.azure_endpoint,
+                azure_ad_token_provider=self.azure_ad_token_provider,
+                api_key=self.api_key,
+                version=self.api_version,
+                azure_deployment=self.azure_deployment,
+                azure_ad_token=self.azure_ad_token,
+                organization=self.organization,
+                max_retries=self.max_retries,
+                **self.extra_kwargs,
+            )
+
+        else:
+            raise ValueError(f"Invalid client_type {self.client_type}")
+
+
+# Allow rough backwards compatibility
+OpenAIClientParameters = OpenAIClientWrapper
+
+
+class OpenAI(LLM):
+    def __init__(
+        self,
+        model_name: Union[OpenAIModels, OpenAIModel, str],
+        api_key=None,
+        client_wrapper: Optional[OpenAIClientWrapper] = None,
+        params: Optional[OpenAIClientParameters] = None,
+        **kwargs,
+    ):
+        if isinstance(model_name, OpenAIModels):
+            self.model = model_name.value
+        elif isinstance(model_name, OpenAIModel):
+            self.model = model_name
+        elif isinstance(model_name, str):
+            self.model = OpenAIModels.from_name(model_name).value
+
+        if self.model.name == OpenAIModels.TEXT_DAVINCI.value.name:
+            logger.warn("text-davinci-003 is deprecated. Falling back to gpt-3.5-turbo-instruct")
+            self.model = OpenAIModels.GPT_3_5_TURBO_INSTRUCT.value
+
+        self.model_name = self.model.name
+        super().__init__(self.model_name)
+
+        # This is somewhat complex to provide a degree of backward compatibility.
+        if client_wrapper is None:
+            if params is not None:
+                client_wrapper = params
+            else:
+                if api_key is not None:
+                    kwargs.update({"api_key": api_key})
+
+                client_wrapper = OpenAIClientWrapper(**kwargs)
+
+        else:
+            if api_key is not None:
+                client_wrapper.api_key = api_key
+
+        self.client_wrapper = client_wrapper
+        self._client = self.client_wrapper.get_client()
+
+    # The actual openai client is not pickleable, This just says to pickle the wrapper, which can be used to
+    # recreate the client on the other end.
+    def __reduce__(self):
+        def deserializer(kwargs):
+            return OpenAI(**kwargs)
+
+        kwargs = {"client_wrapper": self.client_wrapper, "model_name": self.model_name}
+
+        return (deserializer, (kwargs,))
+
+    def is_chat_mode(self):
+        return self.model.is_chat
+
+    def generate(self, *, prompt_kwargs: dict, llm_kwargs: Optional[dict] = None) -> Any:
+        if llm_kwargs is not None:
```

Comment:
this method always confused me, can we have an explicit parameter that says use guidance vs not? do we always need llm_kwargs to make a direct open ai call?

Reply:
Yes, I agree it's weird. Let me play around with that in a separate PR to avoid too much churn.